presto-spi/src/main/java/com/facebook/presto/spi/type/UnscaledDecimal128Arithmetic.java

```diff
@@ -0,0 +1,762 @@
+        long[] longs = new long[NUMBER_OF_LONGS];
+        if (unscaledValue < 0) {
+            longs[0] = -unscaledValue;
+            longs[1] = SIGN_LONG_MASK;
+        }
+        else {
+            longs[0] = unscaledValue;
+        }
+        return Slices.wrappedLongArray(longs);
+    }
+
+    public static BigInteger unscaledDecimalToBigInteger(Slice decimal)
+    {
+        byte[] bytes = decimal.getBytes(0, UNSCALED_DECIMAL_128_SLICE_LENGTH);
+        // convert to big-endian order
+        reverse(bytes);
+        bytes[0] &= ~SIGN_BYTE_MASK;
+        return new BigInteger(isNegative(decimal) ? -1 : 1, bytes);
+    }
+
+    public static long unscaledDecimalToUnscaledLong(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        long hi = getLong(decimal, 1);
+        boolean negative = isNegative(decimal);
+
+        if (hi != 0 || ((lo > Long.MIN_VALUE || !negative) && lo < 0)) {
+            throwOverflowException();
+        }
+
+        return negative ? -lo : lo;
+    }
+
+    public static long unscaledDecimalToUnscaledLongUnsafe(Slice decimal)
+    {
+        long lo = getLong(decimal, 0);
+        return isNegative(decimal) ? -lo : lo;
+    }
+
+    public static Slice rescale(Slice decimal, int rescaleFactor)
+    {
+        if (rescaleFactor == 0) {
+            return decimal;
+        }
+        else {
+            Slice result = unscaledDecimal();
+            rescale(decimal, rescaleFactor, result);
+            return result;
+        }
+    }
+
+    public static void rescale(Slice decimal, int rescaleFactor, Slice result)
+    {
+        if (rescaleFactor == 0) {
+            copyUnscaledDecimal(decimal, result);
+        }
+        else if (rescaleFactor > 0) {
+            if (rescaleFactor >= POWERS_OF_TEN.length) {
+                throwOverflowException();
+            }
+            multiply(decimal, POWERS_OF_TEN[rescaleFactor], result);
+        }
+        else {
+            scaleDownRoundUp(decimal, -rescaleFactor, result);
+        }
+    }
+
+    public static Slice multiply(Slice left, Slice right)
+    {
+        Slice result = unscaledDecimal();
+        multiply(left, right, result);
+        return result;
+    }
+
+    public static void multiply(Slice left, Slice right, Slice result)
+    {
+        int l0 = getInt(left, 0);
+        int l1 = getInt(left, 1);
+        int l2 = getInt(left, 2);
+        int l3 = getInt(left, 3);
+
+        int r0 = getInt(right, 0);
+        int r1 = getInt(right, 1);
+        int r2 = getInt(right, 2);
+        int r3 = getInt(right, 3);
+
+        // the combinations below definitely result in an overflow
+        if (((r3 != 0 && (l3 | l2 | l1) != 0) || (r2 != 0 && (l3 | l2) != 0) || (r1 != 0 && l3 != 0))) {
+            throwOverflowException();
+        }
+
+        int z0 = 0;
+        int z1 = 0;
+        int z2 = 0;
+        int z3 = 0;
+        int z4 = 0;
+        int z5 = 0;
+        int z6 = 0;
+        int z7 = 0;
+
+        if (l0 != 0) {
```

Comment:
The thing is that the same block is copied 4 times, with just indexes changes. We had to copy it due to lack of inline arrays in java. In order to make it more readable I've decided to keep the same pattern for all the copies, so, it is intuitive that the code is the same.